---
title: "Accessing HapMap Data from biomaRt"
author: "Bruce Cochrane"
date: "`r format(Sys.Date(),format="%B %d %Y")`"
output: html_document
---

Introduction
=========
In this vignette, we will use the package biomaRt to access genotype data from [HapMap](http://www.hapmap.org).  We will do the following

1.  Construct a biomaRt query to retrieve genotype data for a random sample of SNPs from the human genome.
2.  Convert those data to a genind object
3.  Analyze the ouptut 

Assumptions
============
- Assumption 1
- Assumption 2

Data
==========
Links to information about the data if needed e.g. package vignettes.    
Link to the data to download for the workflow if it is public data or in this repository. Can download into R here and show the code.   
Format the users data needs to be in for this to work (import data and show a section to demonstrate if needed).  

Resources/Packages required
=============
Packages required. Say where these can be found, can link to the list of packages on our repository page here also.   

Loading the required packages:  
```{r, packages, message=FALSE}
library(hierfstat)
library(adegenet)
library(RCurl)
library(biomaRt)
```

Analysis divided into sections  
==========================================
Example Sections
##Section 1: Load the data

First, we need our list of SNPs.  The following command will retrieve a list of 500 SNP names randomly selected from the more than 4 million ones in the HapMap database 

```{r, loaddata, eval=FALSE}
snps <-read.table(textConnection(getURL("https://dl.dropboxusercontent.com/u/9752688/qpopgen/Summer15Development/DATA/randomsnps.txt")))
```
###Section 2.  Constructing a query

A Biomart query has three components:

1.  A **mart** ,or data source
2.  A set of **attributes**, the information we want to return in each record
3.  **Filters**, which specify which records we wish to return. In this case, we will use the random snps we have loaded.

Note that you can learn much more about this with the command

    vignette("biomaRt")
    
In our case, we will set up the database and specify attributes as follows


```{r}
mart <-useMart("HapMap_rel27",dataset="hm27_variation") #select HapMap release 27 data for all populations
att <-c("sample_id","chrom","start", "pop_id","pop_code_genotype","marker1","genotype") #select filters
flt <-"marker_name"
```

And our query is now 
```{r}
dat <-getBM(attributes=att,filters=flt,values=snps,mart=mart)
```

##Section 3: Processing the data

The data in their present form are rather messy; the following steps will clean it up and make it more workable

### Splitting the data bay population

```{r}
dat.spl <-split(dat,dat$pop_code_genotype)
```

### Creating a list of Genind objects

The following code will create a function that will convert the data into a list of genind objects:
```{r}
bmhap2df <-function(df1){
  indiv <-unlist(strsplit(df1[1,1]," ")) # extract names of individuals
  snp.ids <-df1$marker1 # extract snp names
  genos <-df1$genotype # create a variable of the individual genotypes
  genos.spl <-sapply(genos,strsplit," ") # separate multilocus genotypes into single loci
  genos.df <-do.call(cbind.data.frame,genos.spl) #convert this to a data frame
  genos.df[genos.df=="NN"] <-NA #convert missing data to NA
  rownames(genos.df) <-indiv # name rows
  colnames(genos.df) <-snp.ids # and columns
  return(genos.df)
  
}
```
And now apply the function to our data:

```{r}
dat.gen <-lapply(dat.spl,bmhap2df)
```

At this point, the population-specific data sets can be analyzed individually; i. e.

```{r}
names(dat.gen)
chb.gen <- df2genind(dat.gen$ASW)
summary(chb.gen)
```

### Creating a multipopulation Genind object

More usefully, perhaps, we could create a single genind object containing data for all 11 populations.  To do so, we have to do the following:

1.  Identify those snps for which there are genotypes in all eleven data sets.
2.  Combine them into a single data frame.
3.  Create a vector of population names
4.  Convert the data frame to a genind object.

Those steps can be performed as follows:

#### Identify common SNPS
```{r}
snps.all <-lapply(dat.gen,colnames)
snps.common <-Reduce(intersect,(snps.all))
length(snps.common)
```
#### Create a combined data frame with data for those snps
```{r}
dat.common <-lapply(dat.gen, function (x) x[,snps.common])
dat.comb <-do.call(rbind.data.frame, dat.common)
dim(dat.comb)
```
And now we have a database of 1207 samples, with 139 snps each.

####  Create a single genind object

The data frame created above has all of the necessary genotype information.  We also need a vector of population sources (one for each sample.)  

```{r}
pops <-names(dat.common)
popsize <-sapply(dat.common,dim)[1,]
npops <-length(dat.common)
popvec <-unlist(lapply(1:npops, function (x) rep(pops[x],popsize[x])))
```

Now we create and summarize the genind object:

```{r}
dat.comb.gen <-df2genind(dat.comb,pop=popvec)
summary(dat.comb.gen)
```

##Analysis

At this point, we are following the lead of Stephanie Manel in her "Calculating Basic Population Statistics" vignette.  For example

### Calculation of F statistics

We can use the hierfstat function basic.stats for this:

```{r}
dat.comb.hf <-genind2hierfstat(dat.comb.gen)
fstat.basic <-basic.stats(dat.comb.hf,digits=2)
fstat.basic$overall
```

### Principal components

```{r}
x <-indpca(dat.comb.hf,ind.labels=popvec)
plot(x,cex=.5,col=c(1:npops))
```

### kmeans clusterig
```{r}
dat.clu <-find.clusters(dat.comb.gen,n.pca=3)
```
```{r}
plot(dat.clu$Kstat,type="b")
```
### Look at snp distribution

(may want to put this earlier) 
```{r}
barplot(table(dat$chrom),las=2)
```

### Plot Fst by position

```{r}
par(mfrow=c(5,5))

```

  
##Section N: Conclusions drawn from the analysis  

Whats next
=======================
Information on further analysis that could be done, other workflows 

###Session Information

```{r,sessioninfo, echo=FALSE}
sessionInfo()
```
### To be deleted
```{r}
dat <-load("./DATA/allsnps.txt")
```
sample 500
```{r}
random.snps <-sample(snp.list$marker1,500)
```
Try a quick query
```{r}
mart.all <-useMart("HapMap_rel27",dataset="hm27_variation")
all.dat <-getBM(attributes=c("sample_id","pop_id","pop_code_genotype","marker1","genotype"),filters="marker_name",values=random.snps,mart=mart.all)
```
```{r}
write.table(random.snps,file="./DATA/randomsnps.txt")
```
This will be a long shot
```{r}
par(mfrow=c(5,5))
dat.by.snp<-split(dat,dat$marker1)
dat.by.snp <-dat.by.snp[snps.common]
```
This is going to be really hard.  what we need to do is to create a matrix consisting of the chromosome number, location, snp name and Fst.

```{r}
chrom.number <-sapply (dat.by.snp, function (x) x$chrom[1])
chrom.pos <-sapply(dat.by.snp, function (x) x$start[1])
```

```{r}
fst <- fstat.basic$perloc$Fst
```
OK.  Now cbind the whole mess
```{r}
dat.chrom <-cbind.data.frame(chrom.number,chrom.pos,fst,row.names=snps.common)
```
OK.  Now to do a tapply and plot
```{r}
dat.chrom <-split(dat.chrom,dat.chrom$chrom.number)
```
OK!  Now lapply a plot

```{r}
sapply(dat.chrom, function(x) plot(x$chrom.pos,x$fst,xlab=NULL, main=x$chrom.number[1],ylab="Fst"))
```
So the above is not all that useful.  Quit for now
